src/main/java/frc/robot/subsystems/vision/VisionIOPhotonLib.java

```diff
@@ -0,0 +1,197 @@
+package frc.robot.subsystems.vision;
+
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Optional;
+
+import org.photonvision.EstimatedRobotPose;
+import org.photonvision.PhotonCamera;
+import org.photonvision.PhotonPoseEstimator;
+import org.photonvision.PhotonPoseEstimator.PoseStrategy;
+import org.photonvision.targeting.PhotonPipelineResult;
+import org.photonvision.targeting.PhotonTrackedTarget;
+
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
+import edu.wpi.first.apriltag.AprilTagFields;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Transform3d;
+import edu.wpi.first.networktables.NetworkTableEvent;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import frc.robot.Constants;
+import frc.robot.subsystems.drive.Drive;
+
+public class VisionIOPhotonLib implements VisionIO
+{
+    private final PhotonCamera  _camera;
+    private double              _captureTimesStamp = 0.0;
+    private Pose2d              _pose              = new Pose2d();
+    private boolean             _hasPose           = false;
+    private double[]            _distances         = new double[] {};
+    private int[]               _targetIds         = new int[] {};
+    private double[]            _targetYaws        = new double[] {};
+    private int                 _numProcessedTargets;
+    private AprilTagFieldLayout fieldLayout;
+    private PhotonPoseEstimator _poseEstimator;
+
+    public VisionIOPhotonLib(Drive drive)
+    {
+        _camera = new PhotonCamera(Constants.Vision.PHOTON_CAMERA_NAME);
+
+        try
+        {
+            // Load the 2024 field layout
+            fieldLayout = AprilTagFieldLayout.loadFromResource(AprilTagFields.k2024Crescendo.m_resourceFile);
+        }
+        catch (Exception e)
+        {
+            System.err.println("Failed to load AprilTag layout: " + e.getMessage());
+            throw new RuntimeException(e);
+        }
+
+        _poseEstimator = new PhotonPoseEstimator(fieldLayout, PoseStrategy.MULTI_TAG_PNP_ON_COPROCESSOR, Constants.Vision.ROBOT_TO_CAMERA);
+        _poseEstimator.setMultiTagFallbackStrategy(PoseStrategy.LOWEST_AMBIGUITY);
+
+        NetworkTableInstance.getDefault().addListener(NetworkTableInstance.getDefault().getEntry("/photonvision/" + Constants.Vision.PHOTON_CAMERA_NAME + "/latencyMillis"), EnumSet.of(NetworkTableEvent.Kind.kValueRemote), event ->
+        {
+            var results = _camera.getAllUnreadResults();
+            if (results.isEmpty())
+            {
+                return;
+            }
+            // get the latest result (last in the list)
+            var result = results.get(results.size() - 1);
```

Comment:
I'm assuming this gets whatever the latest result is in case there are multiple results available? We could probably think about changing this to use the multithreaded "multi-input" strategy the drivebase uses for odometry with its queues. That might work here as well and could possibly provide more vision data to the odometer.